File: demos/cpp/url-monitor-cli/steps/step-5-implement-phase1/src/checker.cpp
#include "checker.h"
#include <cpr/cpr.h>

CheckResult check_url(const std::string& url, int timeout_seconds) {
    CheckResult result;
    auto response = cpr::Get(
        cpr::Url{url},
        cpr::Timeout{timeout_seconds * 1000},
        cpr::Redirect(10L));

    result.elapsed_ms = static_cast<long>(response.elapsed * 1000);

    if (response.error.code != cpr::ErrorCode::OK) {
        result.success = false;
        result.error = response.error.message;
        return result;
    }

    result.status_code = response.status_code;
    result.success = (response.status_code < 400);
    if (!result.success) {
        result.error = "HTTP " + std::to_string(response.status_code);
    }
    return result;
}

File: solutions/csharp/url-monitor-cli/src/UrlMonitor/Checker.cs
using System.Diagnostics;

namespace UrlMonitor;

public enum Status { Unknown, Up, Down }

public record CheckResult(int HttpStatus, string? Error, double TotalMs);

public static class Checker
{
    /// <summary>
    /// Perform an HTTP GET check against the given URL.
    /// </summary>
    public static async Task<CheckResult> Check(HttpClient client, string url, int timeoutSeconds)
    {
        var sw = Stopwatch.StartNew();

        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutSeconds));
            var response = await client.GetAsync(url, cts.Token);
            sw.Stop();

            return new CheckResult((int)response.StatusCode, null, sw.Elapsed.TotalMilliseconds);
        }
        catch (TaskCanceledException)
        {
            sw.Stop();
            return new CheckResult(0, "timeout", sw.Elapsed.TotalMilliseconds);
        }
        catch (HttpRequestException hre)
        {
            sw.Stop();
            var errorType = ClassifyHttpError(hre);
            return new CheckResult(0, errorType, sw.Elapsed.TotalMilliseconds);
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new CheckResult(0, ex.Message, sw.Elapsed.TotalMilliseconds);
        }
    }

    /// <summary>
    /// Classify a check result into Up or Down.
    /// Up only if no error AND HttpStatus == 200.
    /// </summary>
    public static Status Classify(CheckResult result)
    {
        if (result.Error is not null)
            return Status.Down;

        return result.HttpStatus == 200 ? Status.Up : Status.Down;
    }

    /// <summary>
    /// Return the lowercase string name for a status.
    /// </summary>
    public static string StatusName(Status status) => status switch
    {
        Status.Unknown => "unknown",
        Status.Up => "up",
        Status.Down => "down",
        _ => "unknown"
    };

    private static string ClassifyHttpError(HttpRequestException ex)
    {
        var message = ex.Message.ToLowerInvariant();

        if (message.Contains("connection refused") || message.Contains("actively refused"))
            return "connection_refused";
        if (message.Contains("name resolution") || message.Contains("no such host"))
            return "dns_resolution_failed";
        if (message.Contains("ssl") || message.Contains("tls") || message.Contains("certificate"))
            return "ssl_error";

        return $"http_error: {ex.Message}";
    }
}
